newton/_src/utils/recorder.py

```diff
@@ -34,6 +35,196 @@
+        self._buffer.clear()
+        self._start = 0
+        self._size = 0
+
+    def to_list(self) -> list[T]:
+        """Convert the ring buffer to a regular list."""
+        return [self[i] for i in range(self._size)]
+
+    def from_list(self, items: list[T]) -> None:
+        """Replace buffer contents with items from a list."""
+        self.clear()
+        for item in items:
+            self.append(item)
+
+
+class ArrayCache(Generic[T]):
+    """
+    Cache that assigns a monotonically increasing index to each unique key and stores an object with it.
+
+    - Keys are uint64-compatible integers (use Python int).
+    - Values are stored alongside the assigned index.
+    - During serialization, repeated keys return their existing index; new keys return -1 and are added.
+    - During deserialization, lookups happen by index and return the associated object or raise if missing.
+    """
+
+    def __init__(self):
+        self._key_to_entry: dict[int, tuple[int, T]] = {}
+        self._index_to_entry: dict[int, T] = {}
+        self._next_index: int = 1
+
+    def try_register_pointer_and_value(self, key: int, value: T) -> int:
+        """
+        Register an object under a numeric key.
+
+        Args:
+            key: Unsigned 64-bit compatible integer key
+            value: Object to cache
+
+        Returns:
+            Existing index if the key already exists; otherwise 0 after inserting a new entry.
+        """
+        if key in self._key_to_entry:
+            existing_index, _ = self._key_to_entry[key]
+            return existing_index
+
+        assigned_index = self._next_index
+        self._next_index += 1
+        self._key_to_entry[key] = (assigned_index, value)
+        self._index_to_entry[assigned_index] = value
+        return 0
+
+    def try_get_value(self, index: int) -> T:
+        """
+        Resolve an object by its index.
+
+        Args:
+            index: Previously assigned index
+
+        Returns:
+            The object associated with the given index.
+
+        Raises:
+            KeyError if the index cannot be found.
+        """
+        if index in self._index_to_entry:
+            return self._index_to_entry[index]
+        raise KeyError(f"ArrayCache: index {index} not found")
+
+    def try_register_pointer_and_value_and_index(self, key: int, value: T, index: int) -> int:
+        """
+        Register an object with an explicit, well-defined index (used during deserialization).
+
+        - If the key already exists, the stored index must equal the provided index.
+          Returns that index, or raises on mismatch.
+        - If the key is new, the provided index must not be used by another entry.
+          Adds the mapping and returns the index.
+        - Advances the internal next-index counter if necessary.
+        """
+        if key in self._key_to_entry:
+            existing_index, existing_value = self._key_to_entry[key]
```

Comment:
It would be good to avoid the repeated query for a key in a dictionary and then using the result of the query in a second call.
Instead you can do
```py
existing_index, existing_value = self._key_to_entry.get(key, (None, None))
if existing_index is not None:
    ...
```
This will be faster than having two separate queries. I see these constructs a couple times below in the code, so would be good to get those replaced by `dict.get(key, default_value)`.